File: frontend/src/types/task.ts
export type TaskStatus = "PENDING" | "PROCESSING" | "COMPLETED" | "FAILED";

export type TaskEventType =
  | "CREATED"
  | "PROCESSING_STARTED"
  | "COMPLETED"
  | "FAILED"
  | "RETRIED";

export interface Task {
  id: string;
  title: string;
  description: string | null;
  status: TaskStatus;
  retry_count: number;
  error_message: string | null;
  created_at: string;
  updated_at: string;
}

export interface TaskEvent {
  id: number;
  event_type: TaskEventType;
  created_at: string;
  metadata: Record<string, unknown> | null;
}

export interface TaskDetail extends Task {
  events: TaskEvent[];
}

export interface TaskCreate {
  title: string;
  description?: string;
}

export interface TaskFilters {
  status?: TaskStatus;
  search?: string;
}

export interface TaskList {
  items: Task[];
  total: number;
}

export interface Metrics {
  total: number;
  by_status: Record<TaskStatus, number>;
  queue_length: number;
}

File: backend/app/schemas/task.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.domain.task import TaskEventType, TaskStatus


class TaskCreate(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    description: str | None = None


class TaskRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str
    description: str | None
    status: TaskStatus
    retry_count: int
    error_message: str | None
    created_at: datetime
    updated_at: datetime


class TaskEventRead(BaseModel):
    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    id: uuid.UUID
    event_type: TaskEventType
    created_at: datetime
    metadata: dict | None = Field(default=None, validation_alias="metadata_")


class TaskDetail(TaskRead):
    events: list[TaskEventRead]


class TaskList(BaseModel):
    items: list[TaskRead]
    total: int


class Metrics(BaseModel):
    total: int
    by_status: dict[TaskStatus, int]
    queue_length: int
